pkg/packetfilter/packetfilter.go

```diff
@@ -0,0 +1,208 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package packetfilter
+
+import (
+	"github.com/pkg/errors"
+)
+
+const (
+	TableTypeFilter TableType = iota
+	TableTypeRoute            // mangle
+	TableTypeNAT
+	TableTypeMAX
+)
+
+type TableType uint32
+
+type GlobalEgressHandle string
+
+type RuleAction uint32
+
+const (
+	RuleActionJump RuleAction = iota
+	RuleActionAccept
+	RuleActionMss
+	RuleActionMark
+	RuleActionSNAT
+	RuleActionDNAT
+	RuleActionMAX
+)
+
+type RuleProto uint32
+
+const (
+	RuleProtoUndefined RuleProto = iota
+	RuleProtoAll
+	RuleProtoTCP
+	RuleProtoUDP
+	RuleProtoICMP
+)
+
+type MssClampType uint32
+
+const (
+	UndefinedMSS MssClampType = iota
+	ToPMTU
+	ToValue
+)
+
+type ChainHook uint32
+
+const (
+	ChainHookPrerouting ChainHook = iota
+	ChainHookInput
+	ChainHookForward
+	ChainHookOutput
+	ChainHookPostrouting
+	ChainHookMAX
+)
+
+type ChainPriority uint32
+
+const (
+	ChainPriorityFirst ChainPriority = iota
+	ChainPriorityLast
+)
+
+type ChainType uint32
+
+const (
+	ChainTypeFilter ChainType = iota
+	ChainTypeRoute            // mangle
+	ChainTypeNAT
+	ChainTypeMAX
+)
+
+type ChainPolicy uint32
+
+const (
+	ChainPolicyAccept ChainType = iota
+	ChainPolicyDrop
+)
+
+type Rule struct {
+	DestCIDR string
+	SrcCIDR  string
+
+	SrcSetName  string
+	DestSetName string
+
+	SnatCIDR string
+	DnatCIDR string
+
+	OutInterface string
+	InInterface  string
+	TargetChain  string
+	MssValue     string
+
+	DPort     string
+	MarkValue string
+	Action    RuleAction
+	Proto     RuleProto
+	ClampType MssClampType
+}
+
+// Supported policy values are accept (which is the default) or drop.
+type ChainRegular struct {
+	Name   string
+	Policy ChainPolicy
+}
+
+type ChainIPHook struct {
+	Name     string
+	Type     ChainType
+	Hook     ChainHook
+	Priority ChainPriority
+	Policy   ChainPolicy
+	JumpRule *Rule
+}
+
+type Interface interface {
+	// Chains
+	ChainExists(table TableType, chain string) (bool, error)
+	CreateIPHookChainIfNotExists(chain *ChainIPHook) error
+	CreateRegularChainIfNotExists(table TableType, chain *ChainRegular) error
+	DeleteIPHookChain(chain *ChainIPHook) error
+	DeleteRegularChain(table TableType, chain string) error
+	ClearChain(table TableType, chain string) error
+
+	// rules
+	Delete(table TableType, chain string, ruleSpec *Rule) error
+	InsertUnique(table TableType, chain string, position int, ruleSpec *Rule) error
+	PrependUnique(table TableType, chain string, ruleSpec *Rule) error
+	UpdateChainRules(table TableType, chain string, rules []*Rule) error
```

Comment:
These methods were originally in the `Adapter`. The iptables lib defines `AppendUnique` so it wasn't in the `Adapter`. Do you know what nftables lib we would use and if it defines similar functions? If not, perhaps we could leave these in the `Adapter` for now? `UpdateChainRules` seems high-level and would probably belong in the `Adapter` to avoid duplication.

Reply:
@yboaron Can you please reply to this comment? Since we don't know yet what the nftables impl will look like, let's keep `InsertUnique`, `PrependUnique`, and `UpdateChainRules` as is in the `Adapter` so we don't have code duplicated between iptables and the fake impl (that was the purpose of the `Adapter`). We can adjust, if needed, when the nftables impl is introduced.

Also `Driver` and `Interface` are backwards, ie `Driver` is the low-level API that the adapter `Interface` builds on.  So `InsertUnique`, `PrependUnique`, and `UpdateChainRules` should be defined in `Interface` and the rest in `Driver`.